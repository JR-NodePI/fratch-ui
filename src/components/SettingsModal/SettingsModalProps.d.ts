import { SettingsModalPosition } from './SettingsModalConstants';

export type SettingsModalProps = {
  className?: string;
  items?: JSX.Element[];
  position?: (typeof SettingsModalPosition)[keyof typeof SettingsModalPosition];
  visible?: boolean;
  onClose?: () => void;
};
