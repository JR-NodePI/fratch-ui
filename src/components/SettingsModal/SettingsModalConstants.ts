export const SettingsModalPosition = {
  LEFT: 'left',
  RIGHT: 'right',
} as const;

export const TIMEOUT_TO_CLOSE = 100;
