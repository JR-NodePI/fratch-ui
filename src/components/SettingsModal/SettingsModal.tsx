import { useEffect, useState } from 'react';

import { c } from '../../helpers/classNameHelpers';
import { isAscendantEvenTargetByID } from '../../helpers/htmlSelectorsHelpers';
import { IconVerticalDots } from '../Icons/Icons';
import { SettingsModalPosition } from './SettingsModalConstants';
import { SettingsModalProps } from './SettingsModalProps';

import styles from './SettingsModal.module.css';

export default function SettingsModal({
  className,
  items = [],
  position = SettingsModalPosition.LEFT,
}: SettingsModalProps): JSX.Element {
  const [openerWrapperId] = useState(crypto.randomUUID());
  const [visible, setVisible] = useState(false);

  useEffect(() => {
    const handleClick = (event: MouseEvent): void => {
      if (!isAscendantEvenTargetByID(event, openerWrapperId)) {
        setVisible(false);
      }
    };
    window.document.body.addEventListener('click', handleClick, true);
    return () => {
      window.document.body.removeEventListener('click', handleClick);
    };
  }, [openerWrapperId]);

  const handleOpenerClick = (): void => {
    setVisible(!visible);
  };

  return (
    <div
      className={c(
        className,
        styles.settings_menu,
        visible ? styles.open : styles.close,
        styles[position]
      )}
      id={openerWrapperId}
    >
      <ul className={c(styles.menu)}>
        {items.map((item, index) => (
          <li key={index}>{item}</li>
        ))}
      </ul>
      <button onClick={handleOpenerClick} className={c(styles.opener)}>
        <IconVerticalDots className={c(styles.opener_icon)} />
      </button>
    </div>
  );
}
