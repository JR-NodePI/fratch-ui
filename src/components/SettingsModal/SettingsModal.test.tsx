import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { describe, expect, it, vi } from 'vitest';

import SettingsModal from './SettingsModal';
import { SettingsModalProps } from './SettingsModalProps';

describe('SettingsModal', () => {
  const settingsMenuId = 'crypto-mock-random-UUID-1234';
  const setup = (
    props: SettingsModalProps = {}
  ): ReturnType<typeof render> & { settingsMenu: Element | null } => {
    const { container, ...data } = render(<SettingsModal {...props} />);
    return {
      container,
      ...data,
      settingsMenu: container.querySelector(`#${settingsMenuId}`),
    };
  };

  beforeEach(() => {
    vi.spyOn(crypto, 'randomUUID').mockReturnValue(settingsMenuId);
  });

  afterEach(() => {
    vi.resetAllMocks();
  });

  it('should render properly', () => {
    const { container } = setup();
    expect(container).toMatchSnapshot();
  });

  it('should render as visible', () => {
    const { container } = setup({ visible: true });
    expect(container).toMatchSnapshot();
  });

  it('should render with items', () => {
    const items = [
      <span>item 1</span>,
      <span>item 2</span>,
      <span>item 3</span>,
    ];
    const { container } = setup({ items });
    expect(container).toMatchSnapshot();
  });

  it.only('should open and then close', async () => {
    const { settingsMenu } = setup();

    await userEvent.click(screen.getByRole('button'));

    expect(settingsMenu?.classList).toMatchInlineSnapshot(`
      DOMTokenList {
        "0": "_settings_menu_21fdeb",
        "1": "_open_21fdeb",
        "2": "_left_21fdeb",
      }
    `);

    await userEvent.tab();

    expect(settingsMenu?.classList).toMatchInlineSnapshot(`
      DOMTokenList {
        "0": "_settings_menu_21fdeb",
        "1": "_open_21fdeb",
        "2": "_left_21fdeb",
      }
    `);
  });

  it('should render with className', () => {
    const { settingsMenu } = setup({ className: 'mock_test_class_name' });

    expect(settingsMenu?.classList).toMatchInlineSnapshot(`
      DOMTokenList {
        "0": "mock_test_class_name",
        "1": "_settings_menu_21fdeb",
        "2": "_close_21fdeb",
        "3": "_left_21fdeb",
      }
    `);
  });

  it('should render with right position', () => {
    const { settingsMenu } = setup({ position: 'right' });

    expect(settingsMenu?.classList).toMatchInlineSnapshot(`
      DOMTokenList {
        "0": "_settings_menu_21fdeb",
        "1": "_close_21fdeb",
        "2": "_right_21fdeb",
      }
    `);
  });
});
