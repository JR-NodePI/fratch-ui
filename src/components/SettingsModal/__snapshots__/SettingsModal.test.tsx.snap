// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SettingsModal > should render properly 1`] = `
<div>
  <div
    class="_settings_menu_21fdeb _close_21fdeb _left_21fdeb"
    id="crypto-mock-random-UUID-1234"
  >
    <ul
      class="_menu_21fdeb"
    />
    <button
      class="_opener_21fdeb"
    >
      <svg
        aria-hidden="true"
        class="_icon_2be8fc _primary_2be8fc _opener_icon_21fdeb"
        fill="currentColor"
        viewBox="0 0 4 15"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M3.5 1.5a1.5 1.5 0 1 1-3 0 1.5 1.5 0 0 1 3 0Zm0 6.041a1.5 1.5 0 1 1-3 0 1.5 1.5 0 0 1 3 0Zm0 5.959a1.5 1.5 0 1 1-3 0 1.5 1.5 0 0 1 3 0Z"
        />
      </svg>
    </button>
  </div>
</div>
`;

exports[`SettingsModal > should render with items 1`] = `
<div>
  <div
    class="_settings_menu_21fdeb _close_21fdeb _left_21fdeb"
    id="crypto-mock-random-UUID-1234"
  >
    <ul
      class="_menu_21fdeb"
    >
      <li>
        <span>
          item 1
        </span>
      </li>
      <li>
        <span>
          item 2
        </span>
      </li>
      <li>
        <span>
          item 3
        </span>
      </li>
    </ul>
    <button
      class="_opener_21fdeb"
    >
      <svg
        aria-hidden="true"
        class="_icon_2be8fc _primary_2be8fc _opener_icon_21fdeb"
        fill="currentColor"
        viewBox="0 0 4 15"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M3.5 1.5a1.5 1.5 0 1 1-3 0 1.5 1.5 0 0 1 3 0Zm0 6.041a1.5 1.5 0 1 1-3 0 1.5 1.5 0 0 1 3 0Zm0 5.959a1.5 1.5 0 1 1-3 0 1.5 1.5 0 0 1 3 0Z"
        />
      </svg>
    </button>
  </div>
</div>
`;
